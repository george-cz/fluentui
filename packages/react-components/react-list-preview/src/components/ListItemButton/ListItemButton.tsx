import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
import { useListItemButtonStyles_unstable } from './useListItemButtonStyles.styles';
import { ButtonProps, renderButton_unstable, useButton_unstable } from '@fluentui/react-button';

export const ListItemButton: ForwardRefComponent<ButtonProps> = React.forwardRef((props, ref) => {
  const state = useButton_unstable({ appearance: 'transparent', ...props }, ref);

  useListItemButtonStyles_unstable(state);
  useCustomStyleHook_unstable('useListItemButtonStyles_unstable')(state);
  return renderButton_unstable(state);
});

ListItemButton.displayName = 'ListItemButton';
