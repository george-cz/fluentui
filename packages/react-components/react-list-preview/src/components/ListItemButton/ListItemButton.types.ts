import { Button } from '@fluentui/react-button';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type ListItemButtonSlots = {
  root: Slot<typeof Button>;
};

/**
 * ListItemButton Props
 */
export type ListItemButtonProps = ComponentProps<ListItemButtonSlots> & {};

/**
 * State used in rendering ListItemButton
 */
export type ListItemButtonState = ComponentState<ListItemButtonSlots>;
