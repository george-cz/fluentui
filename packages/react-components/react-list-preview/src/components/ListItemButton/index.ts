export * from './ListItemButton';
export * from './useListItemButtonStyles.styles';
