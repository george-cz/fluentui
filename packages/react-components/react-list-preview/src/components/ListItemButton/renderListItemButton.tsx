/** @jsxRuntime classic */
/** @jsx createElement */

import { createElement } from '@fluentui/react-jsx-runtime';
import { assertSlots } from '@fluentui/react-utilities';
import type { ListItemButtonState, ListItemButtonSlots } from './ListItemButton.types';

/**
 * Render the final JSX of ListItemButton
 */
export const renderListItemButton_unstable = (state: ListItemButtonState) => {
  assertSlots<ListItemButtonSlots>(state);
  return <state.root />;
};
