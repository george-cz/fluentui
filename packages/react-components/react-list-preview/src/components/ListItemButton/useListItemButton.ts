import * as React from 'react';
import { slot } from '@fluentui/react-utilities';
import type { ListItemButtonProps, ListItemButtonState } from './ListItemButton.types';
import { Button } from '@fluentui/react-button';

/**
 * Create the state required to render ListItemButton.
 *
 * The returned state can be modified with hooks such as useListItemButtonStyles_unstable,
 * before being passed to renderListItemButton_unstable.
 *
 * @param props - props from this instance of ListItemButton
 * @param ref - reference to root HTMLElement of ListItemButton
 */
export const useListItemButton_unstable = (
  props: ListItemButtonProps,
  ref: React.Ref<HTMLElement>,
): ListItemButtonState => {
  return {
    components: {
      root: Button,
    },
    root: slot.always(props, { elementType: Button, defaultProps: { appearance: 'transparent' } }),
  };
};
