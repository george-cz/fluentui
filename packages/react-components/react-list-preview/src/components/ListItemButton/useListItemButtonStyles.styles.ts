import { ButtonState, useButtonStyles_unstable } from '@fluentui/react-button';
import { makeResetStyles, mergeClasses, shorthands } from '@griffel/react';

/**
 * Styles for the root slot
 */
const useStyles = makeResetStyles({
  display: 'block',
  width: '100%',
  textAlign: 'left',
  ...shorthands.padding(0),
});

/**
 * Apply styling to the ListItemButton slots based on the state
 */
export const useListItemButtonStyles_unstable = (state: ButtonState): ButtonState => {
  useButtonStyles_unstable(state);

  const styles = useStyles();
  state.root.className = mergeClasses(state.root.className, styles);

  return state;
};
