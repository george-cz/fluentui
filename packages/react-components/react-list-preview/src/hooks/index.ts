export * from './useListFeatures';
export * from './useListSelection';
