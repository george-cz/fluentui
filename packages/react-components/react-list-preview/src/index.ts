export {
  List,
  listClassNames,
  renderList_unstable,
  useListStyles_unstable,
  useList_unstable,
  ListLayout,
} from './List';

export type { ListProps, ListSlots, ListState, IList } from './List';
export {
  ListItem,
  listItemClassNames,
  renderListItem_unstable,
  useListItemStyles_unstable,
  useListItem_unstable,
} from './ListItem';
export type { ListItemProps, ListItemSlots, ListItemState } from './ListItem';
export {
  ListItemButton,
  listItemButtonClassNames,
  renderListItemButton_unstable,
  useListItemButtonStyles_unstable,
  useListItemButton_unstable,
} from './ListItemButton';
export type { ListItemButtonProps, ListItemButtonSlots, ListItemButtonState } from './ListItemButton';

export { useListFeatures, useListSelection } from './hooks';
