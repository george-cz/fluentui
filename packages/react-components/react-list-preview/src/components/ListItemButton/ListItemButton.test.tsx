import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { ListItemButton } from './ListItemButton';
import { ButtonProps } from '../../../../react-button/src/Button';

describe('ListItemButton', () => {
  isConformant({
    Component: ListItemButton as React.FunctionComponent<ButtonProps>,
    displayName: 'ListItemButton',
    disabledTests: ['component-has-static-classnames-object'],
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<ListItemButton>Default ListItemButton</ListItemButton>);
    expect(result.container).toMatchSnapshot();
  });
});
