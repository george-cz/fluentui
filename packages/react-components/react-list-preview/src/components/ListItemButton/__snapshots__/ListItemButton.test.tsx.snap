// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemButton renders a default state 1`] = `
<div>
  <button
    class="fui-Button"
    type="button"
  >
    Default ListItemButton
  </button>
</div>
`;
